Fix two bugs, add most of the compatibility glue for Solaris 10.

pbhenson opened the pull request:
This is everything needed to compile on Solaris 10 except for handling the missing SO_TIMESTAMP. I'm still working on that, but I thought I'd put this up for review as it might be helpful for AIX or HPUX porting efforts.

I'm not sure how you wanted to handle patches to the patches :), so I just left the original as-is and tagged on the update after it in the same file.

pbhenson committed:
Add paths.h stub for systems (such as Solaris 10) without it.

pbhenson committed:
Add ifaddrs.h stub header file for systems (such as Solaris 10) which
do not have it, and a getifaddrs substitute for Solaris 10.

pbhenson committed:
Add daemon function for Solaris 10 which lacks it.

pbhenson committed:
Add constraint_add function to fix link failure on systems without
libtls.

pbhenson committed:
Update patch to move __progname initialization to before first use to
avoid segfault on systems without native __progname support.

busterb commented on include/paths.h:
```diff
@@ -0,0 +1,18 @@
+#ifndef LIBCOMPAT_PATHS_H
+#define LIBCOMPAT_PATHS_H
+
+#define    _PATH_DEVNULL   "/dev/null"
```
I suspect we could replace this simply with "/dev/null" upstream and remove the need for this.

pbhenson replied:
Works for me. I can't imagine many if any systems targeted by openntpd would have /dev/null someplace else ;).

busterb commented on patches/0009-initialize-setproctitle-where-needed.patch:
```diff
@@ -48,5 +48,39 @@ index c7935bf..44caa80 100644
+ 	struct passwd		*pw;
+ 	extern char		*__progname;
+ 
++	__progname = get_progname(argv[0]);
```
This is already added by a patch in the first place, so I can squash this fix into my patch tree.

pbhenson replied:
Cool. I wasn't sure how you wanted to handle patches to the patches, I could have replaced the whole thing but then the commit id would have changed to me instead of you.

busterb replied:
I have a patches branch that these are derived from here: https://github.com/busterb/openntpd-openbsd/tree/patches

That makes it easier to rebase the patches as needed. I'll just squash your fixes into that and regen the patch files. Thanks!

busterb commented:
Thanks, this looks really good so far. I'm going to be at the OpenBSD hackathon next week, so should have time to coordinate getting this pulled in and getting it further into shape for a full port.

pbhenson commented:
Have fun at the hackathon. So I got it to compile on S10, but my first attempt at working around the missing SO_TIMESTAMP functionality isn't working, it logs:

Jul  5 14:11:13 ike ntpd[10320]: [ID 361431 daemon.crit] recvmsg control format 134.71.247.183: Not owner

Which comes from:

```
    if (T4 < JAN_1970) {
            client_log_error(p, "recvmsg control format", EBADF);
            set_next(p, error_interval());
            return (0);
    }
```

There's probably something else that changed since the update to include SO_TIMESTAMP that's causing the old code to break here, I'll see if I can track it down.

pbhenson committed:
Deal with missing SO_TIMESTAMP by falling back to the previous client.c
implementation when it's not found. SO_TIMESTAMP was added in CVS rev 1.85
by henning@.

busterb commented:
I cherry-picked, merged, or made unnecessary all of these patches, so things should work on Solaris 10 now. Thanks for working on this!
